### DISCUSSION TITLE
Compilation problems

### DISCUSSION THREAD

**Discussion by holyld1987** [gravitational/teleport#17705] [Q&A]
Hi, 
  in addition to making full/all/release for compiling and running, what other ways can you compile and run the source code? I tried the 
  above methods because an error was reported when downloading the package, which made it impossible to compile and run. Thanks!

**Comment by webvictim** [comment]
You can also run things like `make build/tsh`, `make build/teleport`, `make build/tctl` etc to compile individual binaries.

What error are you getting when trying to download? Using precompiled binaries is always going to be easier than compiling yourself where possible.

**Reply by holyld1987** [reply]
I'm new to golang as a Java developer. Can I customize and modify the precompiled binary files? Maybe I seem ignorant when I ask this question. Thank you

**Reply by holyld1987** [reply]
It's true that there are many problems in compiling the source code by myself, which makes it impossible to compile the source code. Of course, it will not run. I tried to compile the source code, but the final result was that there were many timeouts when downloading the package, which led to failure. Now I have a headache. I don't know how to compile and run the source code. I have failed many times......

**Reply by webvictim** [reply]
What do you want to customise and modify about them?

As long as you have Golang installed (i.e. running `go version` gives you a version number rather than an error) you should just be able to `make full` to get the binaries compiled.

(you will need GNU `make` installed as well)

If you have errors, please share what they are and I'll try to assist.

**Reply by holyld1987** [reply]
First of all, thank you for your reply. I'm sorry that this is my first time to contact the Golang language. Please forgive me if there is any improper expression. I tried to download and compile the source code according to the guidance of the document. When I compiled according to the document, many packages failed to download, resulting in compilation failure. Now I want to use this set of open source code to adapt to our own system. The transformation is as follows: 1. Remove login 2. Our system uses the permissions made by keycloak. I don't know whether the open source code of Teleport can integrate keycloak to achieve the unification of the permissions of teleport and keycloak. Maybe my expression is not very good. Please forgive me, thank you.

**Reply by holyld1987** [reply]
Error message：
[root@ecs-350353 teleport]# make full
make[1]: Entering directory '/root/apps/teleport'
echo "init webassets submodule"
init webassets submodule
git submodule update --init webassets
Submodule 'webassets' (https://github.com/gravitational/webassets.git) registered for path 'webassets'
Cloning into '/root/apps/teleport/webassets'...
Submodule path 'webassets': checked out 'e968ecf213676fd5e7400a7c83f6aaebeb56e272'
make[1]: Leaving directory '/root/apps/teleport'
make all WEBASSETS_TAG="webassets_embed"
make[1]: Entering directory '/root/apps/teleport'
VERSION=12.0.0-dev make -f version.mk setver
make[2]: Entering directory '/root/apps/teleport'
for CHART in teleport-cluster teleport-kube-agent teleport-cluster/charts/teleport-operator; do \
	sed -i'.bak' -e "s_^\\.version:\ .*_.version: \\&version \"${VERSION}\"_g" examples/chart/${CHART}/Chart.yaml || exit 1; \
	rm -f examples/chart/${CHART}/Chart.yaml.bak; \
done
make[2]: Leaving directory '/root/apps/teleport'
# Update api module path, but don't fail on error.
make update-api-import-path || true
make[2]: Entering directory '/root/apps/teleport'
go run build.assets/gomod/update-api-import-path/main.go -tags "bpf fips pam desktop_access_rdp linux"
go: downloading github.com/coreos/go-semver v0.3.0
go: downloading github.com/gravitational/trace v1.1.19
go: downloading github.com/gravitational/logrus v1.4.4-0.20210817004754-047e20245621
go: downloading golang.org/x/mod v0.6.0-dev.0.20220419223038-86c51ed26bb4
go: downloading golang.org/x/tools v0.1.12
go: downloading github.com/aws/aws-sdk-go-v2/config v1.17.8
go: downloading github.com/aws/aws-sdk-go-v2 v1.16.16
go: downloading github.com/aws/aws-sdk-go-v2/feature/ec2/imds v1.12.17
go: downloading github.com/ghodss/yaml v1.0.0
go: downloading github.com/gofrs/flock v0.8.1
go: downloading github.com/google/uuid v1.3.0
go: downloading github.com/gravitational/kingpin v2.1.11-0.20220901134012-2a1956e29525+incompatible
go: downloading github.com/jonboulle/clockwork v0.3.0
go: downloading github.com/json-iterator/go v1.1.12
go: downloading github.com/pquerna/otp v1.3.0
go: downloading go.opentelemetry.io/otel/trace v1.11.0
go: downloading go.opentelemetry.io/otel v1.11.0
go: downloading golang.org/x/crypto v0.0.0-20220926161630-eccd6366d1be
go: downloading google.golang.org/grpc v1.50.1
go: downloading k8s.io/apimachinery v0.25.3
go: downloading golang.org/x/net v0.0.0-20221004154528-8021a29435af
go: downloading golang.org/x/sys v0.0.0-20221010170243-090e33056c14
go: downloading github.com/aws/aws-sdk-go-v2/credentials v1.12.21
go: downloading github.com/aws/aws-sdk-go-v2/internal/ini v1.3.24
go: downloading github.com/aws/aws-sdk-go-v2/service/sso v1.11.23
go: downloading github.com/aws/aws-sdk-go-v2/service/ssooidc v1.13.6
go: downloading github.com/aws/aws-sdk-go-v2/service/sts v1.16.19
go: downloading github.com/aws/smithy-go v1.13.3
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/alecthomas/template v0.0.0-20190718012654-fb15b899a751
go: downloading github.com/alecthomas/units v0.0.0-20211218093645-b94a6e3cc137
go: downloading github.com/gravitational/protobuf v1.3.2-0.20201123192827-2b9fcfaffcbf
go: downloading github.com/google/go-cmp v0.5.9
go: downloading github.com/russellhaering/gosaml2 v0.8.1
go: downloading google.golang.org/protobuf v1.28.1
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/modern-go/reflect2 v1.0.2
go: downloading github.com/boombuler/barcode v1.0.1
go: downloading k8s.io/utils v0.0.0-20220728103510-ee6ede2d64ed
go: downloading sigs.k8s.io/yaml v1.3.0
go: downloading golang.org/x/term v0.0.0-20220919170432-7a66f970e087
go: downloading github.com/aws/aws-sdk-go-v2/internal/configsources v1.1.23
go: downloading github.com/aws/aws-sdk-go-v2/service/internal/presigned-url v1.9.17
go: downloading github.com/beevik/etree v1.1.0
go: downloading github.com/mattermost/xml-roundtrip-validator v0.1.0
go: downloading github.com/russellhaering/goxmldsig v1.2.0
go: downloading google.golang.org/genproto v0.0.0-20221010155953-15ba04fc1c0e
go: downloading github.com/golang/protobuf v1.5.2
go: downloading sigs.k8s.io/json v0.0.0-20220713155537-f223a00ba0e2
go: downloading github.com/aws/aws-sdk-go-v2/internal/endpoints/v2 v2.4.17
go: downloading golang.org/x/text v0.4.0
INFO             Using buildFlags: [] update-api-import-path/main.go:51
INFO             the current API version (12.0.0-dev) is not a release or alpha/beta pre-release, continue without updating update-api-import-path/main.go:326
exit status 1
make[2]: *** [Makefile:834: update-api-import-path] Error 1
make[2]: Leaving directory '/root/apps/teleport'
---> Building OSS binaries.
make build/teleport build/tctl build/tsh build/tbot
make[2]: Entering directory '/root/apps/teleport'
GOOS=linux GOARCH=amd64 CGO_ENABLED=1 CGO_LDFLAGS="-Wl,-Bstatic  -Wl,-Bdynamic -Wl,--as-needed" go build -tags "   webassets_embed  " -o build/teleport  -ldflags '-w -s' -trimpath ./tool/teleport
go: downloading github.com/grpc-ecosystem/go-grpc-middleware/providers/openmetrics/v2 v2.0.0-20220714234348-5d0f5fedefc0
go: downloading github.com/grpc-ecosystem/go-grpc-middleware v1.3.0
go: downloading github.com/prometheus/client_golang v1.13.0
go: downloading github.com/gravitational/sftp v1.13.6-0.20220927202521-0e74d42f8055
go: downloading github.com/grpc-ecosystem/go-grpc-middleware/v2 v2.0.0-rc.2.0.20220308023801-e4a6915ea237
go: downloading github.com/beorn7/perks v1.0.1
go: downloading github.com/cespare/xxhash/v2 v2.1.2
go: downloading github.com/prometheus/client_model v0.2.0
go: downloading github.com/prometheus/common v0.37.0
go: downloading github.com/prometheus/procfs v0.8.0
go: downloading github.com/go-ldap/ldap/v3 v3.4.4
go: downloading github.com/aws/aws-sdk-go v1.44.117
go: downloading gopkg.in/square/go-jose.v2 v2.6.0
go: downloading github.com/gravitational/roundtrip v1.0.2
go: downloading go.mongodb.org/mongo-driver v1.10.3
go: downloading github.com/creack/pty v1.1.18
go: downloading github.com/moby/term v0.0.0-20220808134915-39b0c02b01ae
go: downloading github.com/stretchr/testify v1.8.0
go: downloading github.com/kr/fs v0.1.0
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.2
go: downloading github.com/Azure/go-ntlmssp v0.0.0-20220621081337-cb9428e4ac1e
go: downloading github.com/go-asn1-ber/asn1-ber v1.5.4
go: downloading github.com/armon/go-radix v1.0.0
go: downloading github.com/hashicorp/golang-lru v0.5.4
go: downloading github.com/mattn/go-sqlite3 v1.14.15
go: downloading github.com/google/btree v1.1.2
go: downloading github.com/duo-labs/webauthn v0.0.0-20220815211337-00c9fb5711f5
go: downloading github.com/gorilla/websocket v1.5.0
go: downloading go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp v0.36.3
go: downloading golang.org/x/sync v0.0.0-20220929204114-8fcdb60fdcc0
go: downloading k8s.io/client-go v0.25.3
go: downloading github.com/gravitational/oxy v0.0.0-20221006122657-40fb61a9d599
go: downloading github.com/mailgun/timetools v0.0.0-20170619190023-f3a7b8ffff47
go: downloading github.com/mailgun/ttlmap v0.0.0-20170619185759-c1c17f74874f
go: downloading github.com/Azure/azure-sdk-for-go/sdk/azcore v1.1.4
go: downloading github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/redis/armredis/v2 v2.0.0
go: downloading github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/redisenterprise/armredisenterprise v1.0.0
go: downloading github.com/gravitational/go-oidc v0.0.6
go: downloading github.com/gravitational/configure v0.0.0-20180808141939-c3428bd84c23
go: downloading github.com/gravitational/reporting v0.0.0-20210923183620-237377721140
go: downloading github.com/gravitational/predicate v1.2.1
go: downloading go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc v0.36.3
go: downloading github.com/mdlayher/netlink v1.6.2
go: downloading github.com/aws/aws-sdk-go-v2/service/ec2 v1.63.1
go: downloading github.com/gravitational/httprouter v1.3.1-0.20220408074523-c876c5e705a5
go: downloading github.com/sethvargo/go-diceware v0.3.0
go: downloading go.mozilla.org/pkcs7 v0.0.0-20210826202110-33d05740a352
go: downloading go.opentelemetry.io/otel/exporters/otlp/otlptrace v1.11.0
go: downloading go.opentelemetry.io/proto/otlp v0.19.0
go: downloading golang.org/x/oauth2 v0.0.0-20220909003341-f21342109be1
go: downloading google.golang.org/api v0.98.0
go: downloading github.com/ThalesIgnite/crypto11 v1.2.5
go: downloading go.etcd.io/etcd/api/v3 v3.5.5
go: downloading go.etcd.io/etcd/client/v3 v3.5.5
go: downloading cloud.google.com/go/firestore v1.6.1
go: downloading cloud.google.com/go v0.104.0
go: downloading k8s.io/api v0.25.3
go: downloading cloud.google.com/go/iam v0.5.0
go: downloading github.com/Azure/azure-sdk-for-go/sdk/azidentity v1.1.0
go: downloading github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/containerservice/armcontainerservice/v2 v2.1.0
go: downloading github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/mysql/armmysql v1.0.0
go: downloading github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/postgresql/armpostgresql v1.0.0
go: downloading github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/subscription/armsubscription v1.0.0
go: downloading cloud.google.com/go/storage v1.27.0
go: downloading github.com/go-logr/logr v1.2.3
go: downloading github.com/gravitational/ttlmap v0.0.0-20171116003245-91fd36b9004c
go: downloading golang.org/x/exp v0.0.0-20220929160808-de9c53c655b9
go: downloading k8s.io/apiserver v0.25.2
go: downloading k8s.io/klog/v2 v2.80.1
go: downloading go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc v1.11.0
go: downloading go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp v1.11.0
go: downloading go.opentelemetry.io/otel/sdk v1.11.0
go: downloading github.com/alicebob/miniredis/v2 v2.23.0
go: downloading github.com/gravitational/redis/v9 v9.0.0-teleport.1
go: downloading github.com/mailgun/lemma v0.0.0-20170619173223-4214099fb348
go: downloading github.com/golang-jwt/jwt/v4 v4.2.0
go: downloading github.com/golang-jwt/jwt v3.2.2+incompatible
go: downloading github.com/mitchellh/mapstructure v1.5.0
go: downloading github.com/fxamacker/cbor/v2 v2.4.0
go: downloading github.com/flynn/u2f v0.0.0-20180613185708-15554eb68e5d
go: downloading github.com/schollz/progressbar/v3 v3.11.0
go: downloading github.com/felixge/httpsnoop v1.0.3
go: downloading go.opentelemetry.io/otel/metric v0.32.3
go: downloading gopkg.in/mgo.v2 v2.0.0-20190816093944-a6b53ec6cb22
go: downloading github.com/mailgun/minheap v0.0.0-20170619185613-3dbe6c6bf55f
go: downloading github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/compute/armcompute/v3 v3.0.1
go: downloading github.com/jmespath/go-jmespath v0.4.0
go: downloading github.com/josharian/native v1.0.0
go: downloading github.com/mdlayher/socket v0.2.3
go: downloading github.com/coreos/pkg v0.0.0-20180928190104-399ea9e2e55f
go: downloading github.com/grpc-ecosystem/grpc-gateway/v2 v2.11.3
go: downloading github.com/grpc-ecosystem/grpc-gateway v1.16.0
go: downloading cloud.google.com/go/compute v1.10.0
go: downloading github.com/miekg/pkcs11 v1.1.1
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/thales-e-security/pool v0.0.2
go: downloading go.etcd.io/etcd/client/pkg/v3 v3.5.5
go: downloading go.uber.org/zap v1.21.0
go: downloading github.com/googleapis/gax-go/v2 v2.5.1
go: downloading github.com/google/gofuzz v1.2.0
go: downloading github.com/Azure/azure-sdk-for-go/sdk/internal v1.0.0
go: downloading github.com/AzureAD/microsoft-authentication-library-for-go v0.5.1
go: downloading sigs.k8s.io/structured-merge-diff/v4 v4.2.3
go: downloading github.com/moby/spdystream v0.2.0
go: downloading github.com/Azure/go-autorest/autorest v0.11.27
go: downloading github.com/Azure/go-autorest v14.2.0+incompatible
go: downloading github.com/Azure/go-autorest/autorest/adal v0.9.20
go: downloading github.com/imdario/mergo v0.3.13
go: downloading github.com/spf13/pflag v1.0.5
go: downloading go.opentelemetry.io/otel/exporters/otlp/internal/retry v1.11.0
go: downloading github.com/aws/aws-sigv4-auth-cassandra-gocql-driver-plugin v0.0.0-20220331165046-e4d000c0d6a6
go: downloading github.com/gravitational/go-cassandra-native-protocol v0.0.0-20221005103706-b9e66c056e90
go: downloading github.com/gocql/gocql v1.2.1
go: downloading github.com/gravitational/go-mysql v1.5.0-teleport.1
go: downloading github.com/jackc/pgconn v1.13.0
go: downloading github.com/jackc/pgerrcode v0.0.0-20220416144525-469b46aa5efa
go: downloading github.com/elastic/go-elasticsearch/v8 v8.4.0
go: downloading github.com/jackc/pgproto3/v2 v2.3.1
go: downloading github.com/snowflakedb/gosnowflake v1.6.13
go: downloading github.com/gravitational/go-mssqldb v0.11.1-0.20221006130402-25bef12c7ee1
go: downloading github.com/jcmturner/gokrb5/v8 v8.4.3
go: downloading github.com/alicebob/gopher-json v0.0.0-20200520072559-a9ecdc9d1d3a
go: downloading github.com/yuin/gopher-lua v0.0.0-20220504180219-658193537a64
go: downloading github.com/dgryski/go-rendezvous v0.0.0-20200823014737-9f7001d12a5f
go: downloading github.com/pavlo-v-chernykh/keystore-go/v4 v4.4.0
go: downloading github.com/mailgun/metrics v0.0.0-20150124003306-2b3c4565aafd
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/pmezard/go-difflib v1.0.0
go: downloading gopkg.in/yaml.v3 v3.0.1
go: downloading github.com/cloudflare/cfssl v1.6.1
go: downloading github.com/x448/float16 v0.8.4
go: downloading github.com/flynn/hid v0.0.0-20190502022136-f1b9b6cc019a
go: downloading github.com/mattn/go-runewidth v0.0.14
go: downloading github.com/mitchellh/colorstring v0.0.0-20190213212951-d06e56a500db
go: downloading github.com/xeipuuv/gojsonschema v1.2.0
go: downloading github.com/pquerna/cachecontrol v0.1.0
go: downloading go.opencensus.io v0.23.0
go: downloading github.com/coreos/go-systemd/v22 v22.3.2
go: downloading go.uber.org/atomic v1.7.0
go: downloading go.uber.org/multierr v1.7.0
go: downloading golang.org/x/xerrors v0.0.0-20220907171357-04be3eba64a2
go: downloading gopkg.in/inf.v0 v0.9.1
go: downloading k8s.io/kube-openapi v0.0.0-20220928191237-829ce0c27909
go: downloading github.com/google/gnostic v0.6.9
go: downloading golang.org/x/time v0.0.0-20220922220347-f3bd1da661af
go: downloading github.com/kylelemons/godebug v1.1.0
go: downloading github.com/pkg/browser v0.0.0-20210911075715-681adbf594b8
go: downloading github.com/Azure/go-autorest/logger v0.2.1
go: downloading github.com/Azure/go-autorest/tracing v0.6.0
go: downloading github.com/Azure/go-autorest/autorest/date v0.3.0
go: downloading github.com/go-logr/stdr v1.2.2
go: downloading github.com/cenkalti/backoff/v4 v4.1.3
go: downloading github.com/rs/zerolog v1.20.0
go: downloading github.com/golang/snappy v0.0.3
go: downloading github.com/hailocab/go-hostpool v0.0.0-20160125115350-e80d13ce29ed
go: downloading github.com/pingcap/errors v0.11.5-0.20201126102027-b0a155152ca3
go: downloading github.com/siddontang/go-log v0.0.0-20180807004314-8d05993dda07
go: downloading github.com/siddontang/go v0.0.0-20180604090527-bdc77568d726
go: downloading github.com/jackc/chunkreader/v2 v2.0.1
go: downloading github.com/jackc/pgio v1.0.0
go: downloading github.com/jackc/pgpassfile v1.0.0
go: downloading github.com/jackc/pgservicefile v0.0.0-20200714003250-2b9c44734f2b
go: downloading github.com/elastic/elastic-transport-go/v8 v8.1.0
go: downloading github.com/youmark/pkcs8 v0.0.0-20181117223130-1be2e3e5546d
go: downloading github.com/klauspost/compress v1.15.11
go: downloading github.com/xdg-go/scram v1.1.1
go: downloading github.com/xdg-go/stringprep v1.0.3
go: downloading github.com/montanaflynn/stats v0.6.6
go: downloading github.com/Azure/azure-storage-blob-go v0.15.0
go: downloading github.com/apache/arrow/go/arrow v0.0.0-20211112161151-bc219186db40
go: downloading github.com/aws/aws-sdk-go-v2/feature/s3/manager v1.11.33
go: downloading github.com/aws/aws-sdk-go-v2/service/s3 v1.27.11
go: downloading github.com/form3tech-oss/jwt-go v3.2.5+incompatible
go: downloading github.com/gabriel-vasile/mimetype v1.4.1
go: downloading github.com/golang-sql/civil v0.0.0-20190719163853-cb61b32ac6fe
go: downloading github.com/golang-sql/sqlexp v0.0.0-20170517235910-f1bb20e5a188
go: downloading github.com/jcmturner/dnsutils/v2 v2.0.0
go: downloading github.com/jcmturner/gofork v1.7.6
go: downloading github.com/jcmturner/goidentity/v6 v6.0.1
go: downloading gopkg.in/ini.v1 v1.67.0
go: downloading github.com/rivo/uniseg v0.4.2
go: downloading github.com/xeipuuv/gojsonreference v0.0.0-20180127040603-bd5ef7bd5415
go: downloading github.com/googleapis/enterprise-certificate-proxy v0.2.0
go: downloading github.com/golang/groupcache v0.0.0-20210331224755-41bb18bfe9da
go: downloading github.com/pierrec/lz4/v4 v4.1.17
go: downloading github.com/xdg-go/pbkdf2 v1.0.0
go: downloading github.com/Azure/azure-pipeline-go v0.2.3
go: downloading github.com/google/flatbuffers v22.9.29+incompatible
go: downloading github.com/aws/aws-sdk-go-v2/aws/protocol/eventstream v1.4.8
go: downloading github.com/aws/aws-sdk-go-v2/internal/v4a v1.0.14
go: downloading github.com/aws/aws-sdk-go-v2/service/internal/accept-encoding v1.9.9
go: downloading github.com/aws/aws-sdk-go-v2/service/internal/checksum v1.1.18
go: downloading github.com/aws/aws-sdk-go-v2/service/internal/s3shared v1.13.17
go: downloading github.com/hashicorp/go-uuid v1.0.3
go: downloading github.com/google/certificate-transparency-go v1.1.2-0.20210511102531-373a877eec92
go: downloading github.com/xeipuuv/gojsonpointer v0.0.0-20180127040702-4e3ac2762d5f
go: downloading github.com/munnerz/goautoneg v0.0.0-20191010083416-a7dc8b61c822
go: downloading github.com/mattn/go-ieproxy v0.0.9
go: downloading github.com/jcmturner/aescts/v2 v2.0.0
go: downloading github.com/jcmturner/rpc/v2 v2.0.3
go: downloading github.com/emicklei/go-restful/v3 v3.9.0
go: downloading github.com/go-openapi/swag v0.22.3
go: downloading github.com/go-openapi/jsonreference v0.20.0
go: downloading github.com/mailru/easyjson v0.7.7
go: downloading github.com/go-openapi/jsonpointer v0.19.5
go: downloading github.com/josharian/intern v1.0.0
go build github.com/aws/aws-sdk-go-v2/service/ec2: /usr/local/go/pkg/tool/linux_amd64/compile: signal: killed
make[2]: *** [Makefile:278: build/teleport] Error 1
make[2]: Leaving directory '/root/apps/teleport'
make[1]: *** [Makefile:265: all] Error 2
make[1]: Leaving directory '/root/apps/teleport'
make: *** [Makefile:347: full] Error 2
[root@ecs-350353 teleport]# ll

**Reply by webvictim** [reply]
This actually looks fine (despite the errors)

What are the specifications of the machine you're using to compile Teleport?

This message: `/usr/local/go/pkg/tool/linux_amd64/compile: signal: killed` looks to me like maybe the machine is running out of memory (or someone else is killing the process while it's compiling)

**Reply by holyld1987** [reply]
[root@ecs-350353 teleport]# free -kh
              total        used        free      shared  buff/cache   available
Mem:          1.8Gi       112Mi       1.5Gi       9.0Mi       116Mi       1.5Gi
Swap:            0B          0B          0B
This is the memory of my machine. It's too small

**Reply by webvictim** [reply]
Yes. I suspect you'll probably need a minimum of 4GB RAM or so to compile Teleport yourself.

**Reply by holyld1987** [reply]
Thank you for increasing the server memory to 4G, and now the compilation is successful

**Reply by holyld1987** [reply]
I am sorry that I also encountered some problems when compiling the front-end code. Could you please help me check it? Thank you

**Reply by holyld1987** [reply]
Error message:
assets by path /assets/ 3.95 MiB
  assets by path /assets/fonts/ 3.8 MiB
    assets by status 2.36 MiB [big] 8 assets
    assets by path /assets/fonts/*.woff 1.01 MiB 8 assets
    assets by path /assets/fonts/*.woff2 430 KiB 4 assets
  assets by info 157 KiB [immutable] 7 assets
assets by path *.js 2.37 MiB
  asset app.dfce0b35481f359bdda9.js 1.2 MiB [emitted] [immutable] [minimized] [big] (name: app)
  asset vendor.dfce0b35481f359bdda9.js 1.16 MiB [emitted] [immutable] [minimized] [big] (name: vendor) (id hint: defaultVendors) 1 related asset
  asset 52.53944244bee74a659c82.js 4.41 KiB [emitted] [immutable] [minimized]
asset favicon.ico 1.88 KiB [emitted]
asset ../index.html 660 bytes [emitted]
Entrypoint app [big] 2.37 MiB (3.95 MiB) = vendor.dfce0b35481f359bdda9.js 1.16 MiB app.dfce0b35481f359bdda9.js 1.2 MiB 27 auxiliary assets
orphan modules 2.03 MiB [orphan] 831 modules
runtime modules 7.38 KiB 13 modules
cacheable modules 4.34 MiB
  modules by path ../../node_modules/ 2.17 MiB 84 modules
  modules by path ../design/src/ 802 KiB
    modules by path ../design/src/assets/ 605 KiB 31 modules
    modules by path ../design/src/theme/ 23.8 KiB 4 modules
    modules by path ../design/src/Button/ 8.19 KiB 2 modules
    modules by path ../design/src/Icon/ 13.9 KiB 2 modules
    modules by path ../design/src/Menu/ 19.6 KiB 2 modules
  modules by path ./src/ 1.39 MiB 7 modules

WARNING in configuration
The value 'hashed' for option 'optimization.moduleIds' is deprecated. Use 'deterministic' instead.

WARNING in asset size limit: The following asset(s) exceed the recommended size limit (244 KiB).
This can impact web performance.
Assets: 
  /assets/fonts/Ubuntu-Bold.ttf (263 KiB)
  /assets/fonts/Ubuntu-MediumItalic.ttf (301 KiB)
  /assets/fonts/Ubuntu-Regular.ttf (292 KiB)
  /assets/fonts/Ubuntu-BoldItalic.ttf (276 KiB)
  /assets/fonts/Ubuntu-Light.ttf (353 KiB)
  /assets/fonts/Ubuntu-LightItalic.ttf (341 KiB)
  /assets/fonts/Ubuntu-Medium.ttf (277 KiB)
  /assets/fonts/Ubuntu-Italic.ttf (318 KiB)
  app.dfce0b35481f359bdda9.js (1.2 MiB)
  vendor.dfce0b35481f359bdda9.js (1.16 MiB)

WARNING in entrypoint size limit: The following entrypoint(s) combined asset size exceeds the recommended limit (244 KiB). This can impact web performance.
Entrypoints:
  app (2.37 MiB)
      vendor.dfce0b35481f359bdda9.js
      app.dfce0b35481f359bdda9.js

webpack 5.64.4 compiled with 3 warnings in 65254 ms
$ yarn workspace @gravitational/teleport.e build --output-path=../../../dist/e/teleport/app
error Unknown workspace "@gravitational/teleport.e".
info Visit https://yarnpkg.com/en/docs/cli/workspace for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Removing intermediate container 82d647d4521e
The command '/bin/sh -c yarn $NPM_SCRIPT $([ -z $OUTPUT ] || echo --output-path=$OUTPUT)' returned a non-zero code: 1
make[1]: *** [Makefile:37: docker-build] Error 1
make[1]: Leaving directory '/root/apps-web/webapps'
make: *** [Makefile:32: build-teleport] Error 2
